tests/test_binary.nim

```diff
@@ -7,6 +7,260 @@ const
     else: ""
   binaryName = "configlet" & binaryExt
 
+proc cloneExercismRepo(repoName, dest: string; isShallow = false): tuple[
+    output: string; exitCode: int] =
+  # Clones the Exercism repo named `repoName` to the location `dest`.
+  let opts = if isShallow: "--depth 1" else: ""
+  let url = &"https://github.com/exercism/{repoName}/"
+  let cmd = &"git clone {opts} {url} {dest}"
+  result = execCmdEx(cmd)
+
+template execAndCheck(expectedExitCode: int; body: untyped) {.dirty.} =
+  ## Runs `body`, and prints the output if the exit code is non-zero.
+  ## `body` must have the same return type as `execCmdEx`.
+  let (outp, exitCode) = body
+  if exitCode != expectedExitCode:
+    echo outp
+    fail()
+
+func conciseDiff(s: string): string =
+  ## Returns the lines of `s` that begin with a `+` or `-` character.
+  result = newStringOfCap(s.len)
+  for line in s.splitLines():
+    if line.len > 0:
+      if line[0] in {'+', '-'}:
+        result.add line
+        result.add '\n'
+
+proc testsForSync(binaryPath: string) =
+  suite "sync":
+    const psDir = ".test_binary_problem_specifications"
+    const trackDir = ".test_binary_nim_track_repo"
+    removeDir(psDir)
+    removeDir(trackDir)
+
+    test "can clone problem-specifications for test setup":
+      execAndCheck(0):
+        cloneExercismRepo("problem-specifications", psDir)
+
+    test "can clone track repo for test setup":
+      execAndCheck(0):
+        cloneExercismRepo("nim", trackDir)
+
+    test "can set the problem-specifications repo to a known state":
+      execAndCheck(0):
+        execCmdEx(&"git -C {psDir} checkout ef9e1e17c84721ee6e9d5a65c8dd3ba2122eac91")
+
+    test "can set the track repo to a known state":
+      execAndCheck(0):
+        execCmdEx(&"git -C {trackDir} checkout 798a250a5baf44468ff39bf016fafc3c6a5375c2")
```

Comment:
I assume tests are executed in order? I was slightly confused by these tests. It looks like we are testing executing git commands. I assume that you've done this in order to get nice error messages when a step fails?

Reply:
Yes, they're executed in order. I agree they're confusing - it does look like we're just testing `git`.

I'll try to refactor so it's clearer - I didn't love how they looked in the test output either:

```
[Suite] sync
  [OK] can clone problem-specifications for test setup
  [OK] can clone track repo for test setup
  [OK] can set the problem-specifications repo to a known state
  [OK] can set the track repo to a known state
  [OK] `sync --check` exits with 1 and prints the expected output
  [OK] `sync --mode=include` exits with 0 and includes the expected test cases
  [OK] `git diff` shows the expected diff
  [OK] after syncing, another `sync --mode=include` performs no changes
  [OK] after syncing, `sync --check` shows that exercises are up to date
  [OK] the `git diff` output is still the same
```

Probably using one setup step. Would you like to see it in the log like the below, or not at all? 
```
[OK] can prepare a state for a track repo and problem-specifications 
```

But yes, it's possible to have good error messages without showing all the setup steps in the log.

Reply:
> Would you like to see it in the log like the below, or not at all?

My natural tendency would be to omit it from the log as it is not really a test. If something goes wrong, there will be an error message anyway, right?

Reply:
> My natural tendency would be to omit it from the log as it is not really a test. 

Sounds good.

>  If something goes wrong, there will be an error message anyway, right?

Yes.

Reply:
Addressed feedback by replacing `test` with `block`.

In the current implementation, we need the new scope because `execAndCheck` is a `dirty` template so that `outp` is available afterwards, for example here: 
```Nim
    test "`git diff` shows the expected diff":
      execAndCheck(0):
        execCmdEx(diffCmd)

      check outp.conciseDiff() == expectedDiffOutput
```

The test output now look like this: https://github.com/exercism/configlet/runs/1936097992#step:6:158

```
[Suite] sync
  [OK] `sync --check` exits with 1 and prints the expected output
  [OK] `sync --mode=include` exits with 0 and includes the expected test cases
  [OK] `git diff` shows the expected diff
  [OK] after syncing, another `sync --mode=include` performs no changes
  [OK] after syncing, `sync --check` shows that exercises are up to date
  [OK] the `git diff` output is still the same
```

Reply:
Excellent@